src/libutil/subdir-source-accessor.cc

```diff
@@ -0,0 +1,56 @@
+#include "source-accessor.hh"
+
+namespace nix {
+
+struct SubdirSourceAccessor : SourceAccessor
```

Comment:
This can also be done with `FilteringSourceAccessor` since it has a `prefix` field. Or conversely, `FilteringSourceAccessor`'s `prefix` could be removed by using `SubdirSourceAccessor`.

Reply:
Can we do this in a future PR? It wasn't immediately obvious to me how to make that use this (I do think the layering in your second suggestion is right).